trackpad.8
.TH TRACKPAD 8 "24th January 2003"

.SH NAME
trackpad \- tool for changing behaviour of PowerMac trackpads

.SH SYNOPSIS
.B trackpad
[
.B notap
|
.B tap
|
.B drag
|
.B lock
|
.B show
]

.SH DESCRIPTION
.B trackpad
changes the behaviour of the PowerMac trackpad when generating mouse button
events.  The trackpad can be configured to generate a click when tapped,
a click-drag, click-and-lock, or no click at all.

.SH OPTIONS

.TP
.B notap
Don't generate click events when the trackpad is tapped.
.TP
.B tap
Generate a click when the trackpad is tapped.
.TP
.B drag
Generate click-and-hold events, for drag-and-drop, when the trackpad is tapped.
.TP
.B lock
Click and lock when the trackpad is tapped.
.TP
.B show
Show the current state of the trackpad.

.SH AUTHOR
This manual page was written by Jamie Wilkinson <jaq@debian.org> for the
Debian GNU/Linux system (but may be used by others).


debian/changelog
pmac-utils (1.1.3-27+ports) unreleased; urgency=medium

  * Upload to unreleased for powerpc and ppc64.
  * Bump compat level to 9 in debian/compat.
  * Drop #define __USE_GNU from nwnvsetenv.c.
  * Fix syntax error in autoboot.sgml.

 -- John Paul Adrian Glaubitz <glaubitz@physik.fu-berlin.de>  Tue, 08 Jan 2019 17:48:12 +0300

pmac-utils (1.1.3-27) unstable; urgency=medium

  * Incremental packaging updates for Standards-Version: 3.9.7:
    - Record our source format (1.0) in debian/source/format.
    - Stop redundantly setting DH_COMPAT in debian/rules.
    - Add build-indep and build-arch targets to debian/rules.
    - Use dh_prep instead of dh_clean -k in debian/rules.
    - Add comment-only watch file to document upstream state.
    - Use ${misc:Depends} in debian/control for binaries.
    - Remake debian/copyright with a DEP-5 compatible format.

 -- Adam Conrad <adconrad@0c3.net>  Tue, 05 Apr 2016 12:48:28 -0600

pmac-utils (1.1.3-26) unstable; urgency=medium

  * Add myselt to Uploaders with Michael's consent.
  * Rename to pmac-utils, powerpc-utils is another project.
  * Stop building on little-endian arches (LP: #1398163)
  * Drop ancient and obsolete Conflicts/Replaces.

 -- Adam Conrad <adconrad@0c3.net>  Sun, 27 Mar 2016 13:42:07 -0600

powerpc-utils (1.1.3-25) unstable; urgency=medium

  * Update maintainer address. Closes: #762790.
  * Fix compiler warnings.

 -- Michael Schmitz <schmitzmic@gmail.com>  Wed, 29 Oct 2014 20:28:32 +1100

powerpc-utils (1.1.3-24.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Add ppc64el to the Architecture: field. Closes: #738140.

 -- Aurelien Jarno <aurel32@debian.org>  Sat, 20 Sep 2014 22:09:45 +0200

powerpc-utils (1.1.3-24) unstable; urgency=low

  * Fix fixed postinstall handling of adjtime setup (copy /etc/adjtime if it
    exists, create new file with the correct data otherwise). 
    Closes: #528443.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 13 May 2009 20:04:24 +1100

powerpc-utils (1.1.3-23) unstable; urgency=low

  * Fix broken -f option for fnset. Check for presence of ADB devices and bail 
    out on USB based machines.
    Closes: #474814.

  * Fix postinstall handling of adjtime setup (copy /etc/adjtime if it
    exists, create new file otherwise). 
    Closes: #528443.

  * Update bootsched manpage to clarify that it should only be used on 
    Oldworld PowerMacs.
    Closes: #407295.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 13 May 2009 19:34:24 +1100

powerpc-utils (1.1.3-22) unstable; urgency=low

  * Fix problem with nvram reads segfaulting on efika, reported by Robert
    Millan <rm@aybabtu.com> and patched by same.
    Closes: #431116.

  * Honor DEB_BUILD_OPTS=noopt, patch by Robert Millan <rm@aybabtu.com>.
    Closes: #431117.

  * Add Aurelien Gerome <ag@roxor.cx> as co-maintainer.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed,  4 Jul 2007 17:21:40 +0100

powerpc-utils (1.1.3-21) unstable; urgency=low

  * Fix problem with truncated nvram reads/writes, reported by Adam
    Lackorzynski <adam@os.inf.tu-dresden.de>.
    Closes: #413131.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed,  7 Mar 2007 15:57:34 +0100

powerpc-utils (1.1.3-20) unstable; urgency=low

  * Change section to 'utils' to match override file.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Tue, 30 Jun 2006 15:39:34 +0100

powerpc-utils (1.1.3-19) unstable; urgency=low

  * The 'upload lost' release :-(

  * Incorporate newworld autoboot patch by Rich Johnson.

  * Fix description formatting (thanks to Nick Boullis <nboullis@debian.org>
    for the alert).
    Closes: #336972.

  * Add ppc64 arch (just in case biarch powerpc isn't enough).
    Closes: #360797.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Tue, 30 Jun 2006 15:39:34 +0100

powerpc-utils (1.1.3-18) unstable; urgency=low

  * add -D_GNU_SOURCE to Makefile in order to help with 64 bit pointer
    size on ppc64.
    Closes: #326394.

  * Fix problem introduced in -12 (/etc/adjtime converted to symlink 
    into /var fs): Convert symlink to plain copy now.
    Closes: #332926.

  * Fix char signedness bug in nvsetvol (patch by Eugen Dedu
    <Eugen.Dedu@pu-pm.univ-fcomte.fr>.
    Closes: #334547

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 19 Oct 2005 18:07:34 +0200

powerpc-utils (1.1.3-17) unstable; urgency=low

  * Fix for iBook G4 failure to save nvram data (as manifest by failure to
    properly store the boot chime volume). Patch by Joerg Dorchain
    <joerg@dorchain.net>.
    Closes: #321655, #320566.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 08 Aug 2005 14:40:34 +0200

powerpc-utils (1.1.3-16) unstable; urgency=low

  * Brown Paper Bag release: Added bootsched util (for CUDA based Powermacs)
    the source for which I had buried for over three years now.

  * Finally, actually install the autoboot binary (reported missing by
    Herve Eychenne <rv@eychenne.org>).
    Closes: #317021.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Wed, 06 Jul 2005 09:42:44 +0200

powerpc-utils (1.1.3-15) unstable; urgency=low

  * Merge back patches from ubuntu (thanks to Adam Conrad <adconrad@0c3.net>
    for the hint): fix linuxdoc-tools b-d, drop adb_mouse.h header. Update 
    description to include autoboot and lsprop tools.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Mon, 09 May 2005 12:31:24 +0100

powerpc-utils (1.1.3-14) unstable; urgency=low

  * Fixed nvsetvol failure (infinite loop reading NVRAM) on OldWorld
    machines (reported by Christophe Donatsch <cdonatsch@tele2.ch>).
    Closes: #304493.

 -- Michael Schmitz <schmitz@biophys.uni-duesseldorf.de>  Sun, 01 May 2005 11:58:37 +0100

powerpc-utils (1.1.3-13) unstable; urgency=low

  * Fixed nvsetenv's devfsd bug (reported by Roger Leigh <rleigh@debian.org>).
    Closes: #300946.

  * Added autoboot util (for CUDA based Powermacs)


trackpad.c
/* trackpad.c
 *
 * Tool for setting the PowerBook trackpad options on linux
 *
 * bye benh <bh40@calva.net>
 * 2/13/99
 *
 * 3/17/99 - Minor Fix: usage display used to leave /dev/adb open
 *
 * 10/12/99 - addition of 'show' option to aid in pmud debugging
 *
 * Pieces from mousehack, from numerous contributors...
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <fcntl.h>
#include <linux/cuda.h>
#include <linux/adb.h>


#define BUFLEN 80


//#define DEBUG

int fd;

void
send(unsigned char *y, int len)
{
    int n;

#ifdef DEBUG
    printf("send: ");
    for (n=0; n < len; n++)
	printf("0x%02x ",y[n]);
    printf("\n");
#endif

    n = write(fd, y, (size_t) len);
    if (n < len) {
	perror("writing /dev/adb");
	close(fd);
	exit (EXIT_FAILURE);
    }
}

int
listen(unsigned char *y)
{
    int n;
    
    n = read(fd, y, BUFLEN);

#ifdef DEBUG
    printf("listen (%d bytes): ",n);
    if (n >= 0) {
	int i;
	for (i=0; i < n; i++)
	    printf("0x%02x ",y[i]);
    }
    printf("\n");
#endif

    if (n < 0) {
	perror("reading /dev/adb");
	close(fd);
	exit(EXIT_FAILURE);
    }
    
    return n;
}

void
set_program_mode(int id, int set)
{
	unsigned char buf[BUFLEN + 1]; //for offset
	int n;
	
#ifdef DEBUG
	printf("program mode set to: %d\n", set);
#endif
	
	buf[0] = ADB_PACKET;
	buf[1] = ADB_READREG(id, 1);
	send(buf, 2);
	n = listen(buf+1);
	buf[0] = ADB_PACKET;
	buf[1] = ADB_WRITEREG(id, 1);
	buf[8] = set ? 0x0d : 0x03;
	send(buf, 1+n);
	listen(buf);

	buf[0] = ADB_PACKET;
	buf[1] = ADB_READREG(id, 2);
	send(buf, 2);
	n = listen(buf+1);
}

enum
{
	mode_invalid = 0,
	mode_notap,
	mode_tap,
	mode_drag,
	mode_lock,
	mode_show
};


void
set_trackpad(int id, int mode)
{
	unsigned char buf[BUFLEN + 1]; /* for offset */

	set_program_mode(id, 1);
	
#ifdef DEBUG		
	printf("setting mode: %d\n", mode);
#endif

	if (mode == mode_show) {

		printf("READREG(%d, %d) ", id, 2);
		buf[0] = ADB_PACKET;
		buf[1] = ADB_READREG(id, 2);
		send(buf, 2);
		listen(buf);
#ifdef DEBUG
		printf("(reply %d bytes): ",n);
		if (n >= 0) {
			int ii;
			for (ii=0; ii < n; ii++)
				printf("0x%02x ",buf[ii]);
		}
		printf("\n");
#endif
		printf("trackpad settings %stap %sdrag %slock\n",
			buf[1] == 0x99 ? "" : "no",
			buf[2] == 0x94 ? "" : "no",
			buf[4] == 0xff ? "" : "no"
		);
	} else {
		buf[0] = ADB_PACKET;
		buf[1] = ADB_WRITEREG(id, 2);
		buf[2] = (mode < mode_tap) ? 0x19 : 0x99;
		buf[3] = (mode < mode_drag) ? 0x14 : 0x94;
		buf[4] = 0x19;
		buf[5] = (mode < mode_lock) ? 0xb2 : 0xff;
		buf[6] = 0xb2;
		buf[7] = 0x8a;
		buf[8] = 0x1b;
		buf[9] = 0x50;
		send(buf, 10);
		listen(buf);
	
#ifdef DEBUG
		{
		int i, n;
	
		for(i=0; i<4; i++)
		{
			printf("READREG(%d, %d) ", id, i);
			buf[0] = ADB_PACKET;
			buf[1] = ADB_READREG(id, i);
			send(buf, 2);
			n = listen(buf+1);
			printf("(reply %d bytes): ",n);
			if (n >= 0) {
				int ii;
				for (ii=1; ii <= n; ii++)
					printf("0x%02x ",buf[ii]);
			}
			printf("\n");
		}
		}
#endif
	}
	set_program_mode(id, 0);

}

int
locate_trackpad(void)
{
	int i, n;
	
	for (i=1; i<16; i++)
	{
		unsigned char buf[BUFLEN];
		
#ifdef DEBUG		
		printf("testing %d...\n", i);
#endif		
		buf[0] = ADB_PACKET;
		buf[1] = ADB_READREG(i, 1);
		send(buf, 2);
		n = listen(buf);
		if ((n >= 4) && (buf[1] == 0x74) && (buf[2] == 0x70) &&
			(buf[3] == 0x61) && (buf[4] == 0x64))
		{
#ifdef DEBUG		
			printf("found trackpad at %d\n", i);
#endif
			return i;
		}
	}
	return -1;
}

int
main(int argc, char **argv)
{
	int id;
	int mode = mode_invalid;
	
  	if (argc >= 2) {
	  	if (strcmp(argv[1], "notap") == 0)
	  		mode = mode_notap;
	  	else if (strcmp(argv[1], "tap") == 0)
	  		mode = mode_tap;
	  	else if (strcmp(argv[1], "drag") == 0)
	  		mode = mode_drag;
	  	else if (strcmp(argv[1], "lock") == 0)
	  		mode = mode_lock;
	  	else if (strcmp(argv[1], "show") == 0)
	  		mode = mode_show;
	}  	
  	if (mode == mode_invalid)
  	{
  		printf("usage: trackpad notap|tap|drag|lock|show\n");
  		return 0;
  	}
  	
	fd = open("/dev/adb", O_RDWR);
	if (fd < 0) {
		perror("opening /dev/adb");
		exit(EXIT_FAILURE);
	}
  
	id = locate_trackpad();
	if (id < 0)
	{
		printf("no trackpad !\n");
		return 0;
	}
	
	set_trackpad(id, mode);
	
	close(fd);
	return 0;
}

